File: +ccTools/compCustomization.m
function [status, errorMsg] = compCustomization(comp, varargin)

    arguments
        comp {ccTools.validators.mustBeBuiltInComponent}
    end

    arguments (Repeating)
        varargin
    end

    warning('off', 'MATLAB:structOnObject')
    warning('off', 'MATLAB:ui:javaframe:PropertyToBeRemoved')

    status   = true;
    errorMsg = '';

    fHandle  = ancestor(comp, 'figure');
    fWebWin  = struct(struct(struct(fHandle).Controller).PlatformHost).CEF;

    compTag  = struct(struct(struct(comp).Controller)).ViewModel.Id;

    % nargin validation
    if nargin == 1
        error('At least one Name-Value parameters must be passed to the function.')
    elseif mod(nargin-1, 2)
        error('Name-value parameters must be in pairs.')
    end

    % customizations...
    switch class(comp)
    %------------------------- UIFIGURE ----------------------------------%
        case 'matlab.ui.Figure' % OK
            propStruct = InputParser({'windowMinSize'}, varargin{:});

            for ii = 1:numel(propStruct)
                switch propStruct(ii).name
                    case 'windowMinSize'
                        fWebWin.setMinSize(propStruct(ii).value)
                end
            end


    %------------------------ CONTAINERS ---------------------------------%
        case {'matlab.ui.container.ButtonGroup'    ... % OK
              'matlab.ui.container.Panel'          ... % OK
              'matlab.ui.container.CheckBoxTree'   ... % OK
              'matlab.ui.container.Tree'}              % OK
            propStruct = InputParser({'backgroundColor', ...
                                      'borderRadius', 'borderWidth', 'borderColor'}, varargin{:});

            jsCommand = '';
            for ii = 1:numel(propStruct)
                switch propStruct(ii).name
                    case 'backgroundColor'
                        jsCommand = sprintf(['%sdocument.querySelector(''div[data-tag="%s"]'').style.backgroundColor = "transparent";\n' ...
                                               'document.querySelector(''div[data-tag="%s"]'').children[0].style.backgroundColor = "%s";\n'], jsCommand, compTag, compTag, propStruct(ii).value);
                    otherwise
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').children[0].style.%s = "%s";\n', jsCommand, compTag, propStruct(ii).name, propStruct(ii).value);
                end
            end


    %---------------------------------------------------------------------%
        case 'matlab.ui.container.GridLayout' % OK
            propStruct = InputParser({'backgroundColor'}, varargin{:});

            jsCommand = '';
            for ii = 1:numel(propStruct)
                switch propStruct(ii).name
                    case 'backgroundColor'
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').style.backgroundColor = "%s";\n', jsCommand, compTag, propStruct(ii).value);
                end
            end


    %---------------------------------------------------------------------%
        case 'matlab.ui.container.TabGroup' % OK
            propStruct = InputParser({'backgroundColor', 'backgroundHeaderColor',   ...
                                      'borderRadius', 'borderWidth', 'borderColor', ...
                                      'fontFamily', 'fontStyle', 'fontWeight', 'fontSize', 'color'}, varargin{:});

            jsCommand = '';
            for ii = 1:numel(propStruct)
                switch propStruct(ii).name
                    case 'backgroundColor'
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').style.backgroundColor = "transparent";\n', jsCommand, compTag);
                        for jj = 1:numel(comp.Children)
                            compChildrenTag = struct(struct(struct(comp.Children(jj)).Controller)).ViewModel.Id;
                            jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').style.backgroundColor = "%s";\n', jsCommand, compChildrenTag, propStruct(ii).value);
                        end

                    case 'backgroundHeaderColor'
                        jsCommand = sprintf(['%sdocument.querySelector(''div[data-tag="%s"]'').style.backgroundColor = "transparent";\n' ...
                                               'document.querySelector(''div[data-tag="%s"]'').children[1].style.backgroundColor = "%s";\n'], jsCommand, compTag, compTag, propStruct(ii).value);

                    case {'borderRadius', 'borderWidth', 'borderColor'}
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').style.%s = "%s";\n', jsCommand, compTag, propStruct(ii).name, propStruct(ii).value);
                end
            end

            % Font Properties (iterative process, going through all the tabs)
            idx = find(cellfun(@(x) ~isempty(x), cellfun(@(x) find(strcmp({'fontFamily', 'fontStyle', 'fontWeight', 'fontSize', 'color'}, x), 1), {propStruct.name}, 'UniformOutput', false)));
            if ~isempty(idx)
                jsCommand = sprintf(['%svar elements = document.querySelector(''div[data-tag="%s"]'').getElementsByClassName("mwTabLabel");\n' ...
                                       'for (let ii = 1; ii < elements.length; ii++) {\n'], jsCommand, compTag);
                for ll = idx
                    jsCommand = sprintf('%selements[ii].style.%s = "%s";\n', jsCommand, propStruct(ll).name, propStruct(ll).value);
                end
                jsCommand = sprintf('%s}\nelements = undefined;\n', jsCommand);
            end


    %---------------------------------------------------------------------%
        case 'matlab.ui.container.Tab' % OK
            propStruct = InputParser({'backgroundColor'}, varargin{:});

            compParentTag = struct(struct(struct(comp.Parent).Controller)).ViewModel.Id;
            jsCommand = '';
            for ii = 1:numel(propStruct)
                jsCommand = sprintf(['%sdocument.querySelector(''div[data-tag="%s"]'').style.backgroundColor = "transparent";\n', ...
                                       'document.querySelector(''div[data-tag="%s"]'').style.backgroundColor = "%s";\n'], jsCommand, compParentTag, compTag, propStruct(ii).value);
            end


    %-------------------------- CONTROLS ---------------------------------%
        case {'matlab.ui.control.Button'  ...  % OK              
              'matlab.ui.control.ListBox' ...  % OK
              'matlab.ui.control.StateButton'} % OK
            propStruct = InputParser({'backgroundColor', ...
                                      'borderRadius', 'borderWidth', 'borderColor', ...
                                      'fontFamily', 'fontStyle', 'fontWeight', 'fontSize', 'color'}, varargin{:});

            jsCommand = '';
            for ii = 1:numel(propStruct)
                switch propStruct(ii).name
                    case 'backgroundColor'
                        jsCommand = sprintf(['%sdocument.querySelector(''div[data-tag="%s"]'').style.backgroundColor = "transparent";\n' ...
                                               'document.querySelector(''div[data-tag="%s"]'').children[0].style.backgroundColor = "%s";\n'], jsCommand, compTag, compTag, propStruct(ii).value);
                    otherwise
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').children[0].style.%s = "%s";\n', jsCommand, compTag, propStruct(ii).name, propStruct(ii).value);
                end
            end


    %---------------------------------------------------------------------%
        case {'matlab.ui.control.DropDown', ...     % OK
              'matlab.ui.control.EditField' ...     % OK
              'matlab.ui.control.NumericEditField'} % OK
            propStruct = InputParser({'backgroundColor', ...
                                      'borderRadius', 'borderWidth', 'borderColor'}, varargin{:});

            jsCommand = '';
            for ii = 1:numel(propStruct)
                switch propStruct(ii).name
                    case 'backgroundColor'
                        jsCommand = sprintf(['%sdocument.querySelector(''div[data-tag="%s"]'').style.backgroundColor = "transparent";\n' ...
                                               'document.querySelector(''div[data-tag="%s"]'').children[0].style.backgroundColor = "%s";\n'], jsCommand, compTag, compTag, propStruct(ii).value);
                    otherwise
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').children[0].style.%s = "%s";\n', jsCommand, compTag, propStruct(ii).name, propStruct(ii).value);
                end
            end


    %---------------------------------------------------------------------%
        case 'matlab.ui.control.TextArea' % OK
            propStruct = InputParser({'backgroundColor', ...
                                      'borderRadius', 'borderWidth', 'borderColor', ...
                                      'fontFamily', 'fontStyle', 'fontWeight', 'fontSize', 'color', 'textAlign'}, varargin{:});

            jsCommand = '';
            for ii = 1:numel(propStruct)
                switch propStruct(ii).name
                    case 'backgroundColor'
                        jsCommand = sprintf(['%sdocument.querySelector(''div[data-tag="%s"]'').style.backgroundColor = "transparent";\n' ...
                                               'document.querySelector(''div[data-tag="%s"]'').children[0].style.backgroundColor = "%s";\n'], jsCommand, compTag, compTag, propStruct(ii).value);
                    case {'fontFamily', 'fontStyle', 'fontWeight', 'fontSize', 'color', 'textAlign'}
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').getElementsByTagName("textarea")[0].style.%s = "%s";\n', jsCommand, compTag, propStruct(ii).name, propStruct(ii).value);
                    otherwise
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').children[0].style.%s = "%s";\n', jsCommand, compTag, propStruct(ii).name, propStruct(ii).value);
                end
            end


    %---------------------------------------------------------------------%
        case 'matlab.ui.control.Table' % PENDENTE BACKGROUNDCOLOR, BORDERRADIUS, BORDERWIDTH & BORDERCOLOR
            propStruct = InputParser({'backgroundColor', 'backgroundHeaderColor',   ...
                                      'borderRadius', 'borderWidth', 'borderColor', ...
                                      'fontFamily', 'fontStyle', 'fontWeight', 'fontSize', 'color'}, varargin{:});

            jsCommand = '';
            for ii = 1:numel(propStruct)
                switch propStruct(ii).name
                    case 'backgoundColor'
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').children[0].style.backgroundColor = "%s";\n', jsCommand, compTag, propStruct(ii).value);

                    case 'backgroundHeaderColor'
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').getElementsByClassName("mw-table-flex-dynamic-item")[0].style.backgroundColor = "%s";\n', jsCommand, compTag, propStruct(ii).value);

                    case {'borderRadius', 'borderWidth', 'borderColor'}
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').children[0].children[0].style.%s = "%s";\n', jsCommand, compTag, propStruct(ii).name, propStruct(ii).value);
                end
            end

            % Font Properties (iterative process, going through all the columns)
            idx = find(cellfun(@(x) ~isempty(x), cellfun(@(x) find(strcmp({'fontFamily', 'fontStyle', 'fontWeight', 'fontSize', 'color'}, x), 1), {propStruct.name}, 'UniformOutput', false)));
            if ~isempty(idx)
                jsCommand = sprintf(['%svar elements = document.querySelector(''div[data-tag="%s"]'').getElementsByClassName("mw-default-header-cell");\n' ...
                                       'for (let ii = 0; ii < elements.length; ii++) {\n'], jsCommand, compTag);
                for ll = idx
                    jsCommand = sprintf('%selements[ii].style.%s = "%s";\n', jsCommand, propStruct(ll).name, propStruct(ll).value);
                end
                jsCommand = sprintf('%s}\nelements = undefined;\n', jsCommand);
            end


    %---------------------------------------------------------------------%
        case 'matlab.ui.control.CheckBox' % OK
            propStruct = InputParser({'backgroundColor', ...
                                      'borderRadius', 'borderWidth', 'borderColor'}, varargin{:});

            jsCommand = '';
            for ii = 1:numel(propStruct)
                jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').getElementsByClassName("mwCheckBoxRadioIconNode")[0].style.%s = "%s";\n', jsCommand, compTag, propStruct(ii).name, propStruct(ii).value);
            end


    %---------------------------------------------------------------------%
        case 'matlab.ui.control.Slider' % OK
            propStruct = InputParser({'backgroundColor', ...
                                      'trackHeight', 'thumbRotate', 'thumbHeight'}, varargin{:});

            jsCommand = '';
            for ii = 1:numel(propStruct)
                switch propStruct(ii).name
                    case 'backgroundColor'
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').getElementsByClassName("mwSliderTrack")[0].style.backgroundColor = "%s";\n', jsCommand, compTag, propStruct(ii).value);
                    case 'trackHeight'
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').getElementsByClassName("mwSliderTrack")[0].style.height = "%s";\n', jsCommand, compTag, propStruct(ii).value);
                    case 'thumbRotate'
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').getElementsByClassName("mwSliderThumb")[0].style.rotate = "%s";\n', jsCommand, compTag, propStruct(ii).value);
                    case 'thumbHeight'
                        jsCommand = sprintf('%sdocument.querySelector(''div[data-tag="%s"]'').getElementsByClassName("mwSliderThumb")[0].style.height = "%s";\n', jsCommand, compTag, propStruct(ii).value);
                end
            end


    %---------------------------------------------------------------------%
        otherwise
            error('ccTools does not cover the customization of ''%s'' class properties.', class(comp))
    end


    % JS
    try
        if exist('jsCommand', 'var')
            pause(.001)
            fWebWin.executeJS(jsCommand);
        end
        
    catch ME
        status   = false;
        errorMsg = getReport(ME);
    end
end


function propStruct = InputParser(propList, varargin)
    
    p = inputParser;
    d = struct();

    for ii = 1:numel(propList)
        switch(propList{ii})
            % Window
            case 'windowMinSize';         d.windowMinSize   = []; addParameter(p, 'windowMinSize',         d.windowMinSize,   @(x) ccTools.validators.mustBeNumericArray(x, 2, 'NonNegativeInteger'))

            % BackgroundColor
            case 'backgroundColor';       d.backgroundColor = []; addParameter(p, 'backgroundColor',       d.backgroundColor, @(x) ccTools.validators.mustBeColor(x, 'all'))
            case 'backgroundHeaderColor'; d.backgroundColor = []; addParameter(p, 'backgroundHeaderColor', d.backgroundColor, @(x) ccTools.validators.mustBeColor(x, 'all'))

            % Border
            case 'borderRadius';          d.borderRadius    = []; addParameter(p, 'borderRadius',          d.borderRadius,    @(x) ccTools.validators.mustBeCSSProperty(x, 'border-radius'))
            case 'borderWidth';           d.borderWidth     = []; addParameter(p, 'borderWidth',           d.borderWidth,     @(x) ccTools.validators.mustBeCSSProperty(x, 'border-width'))
            case 'borderColor';           d.borderColor     = []; addParameter(p, 'borderColor',           d.borderColor,     @(x) ccTools.validators.mustBeColor(x, 'all'))

            % Font
            case 'textAlign';             d.textAlign       = []; addParameter(p, 'textAlign',             d.textAlign,       @(x) ccTools.validators.mustBeCSSProperty(x, 'text-align'))
            case 'fontFamily';            d.fontFamily      = []; addParameter(p, 'fontFamily',            d.fontFamily,      @(x) ccTools.validators.mustBeCSSProperty(x, 'font-family'))
            case 'fontStyle';             d.fontStyle       = []; addParameter(p, 'fontStyle',             d.fontStyle,       @(x) ccTools.validators.mustBeCSSProperty(x, 'font-style'))
            case 'fontWeight';            d.fontWeight      = []; addParameter(p, 'fontWeight',            d.fontWeight,      @(x) ccTools.validators.mustBeCSSProperty(x, 'font-weight'))
            case 'fontSize';              d.fontSize        = []; addParameter(p, 'fontSize',              d.fontSize,        @(x) ccTools.validators.mustBeCSSProperty(x, 'font-size'))
            case 'color';                 d.color           = []; addParameter(p, 'color',                 d.color,           @(x) ccTools.validators.mustBeColor(x, 'all'))

            % Track and thumb (Slider)
            case 'trackHeight';           d.trackHeight     = []; addParameter(p, 'trackHeight',           d.trackHeight,     @(x) ccTools.validators.mustBeCSSProperty(x, 'height'))
            case 'thumbRotate';           d.thumbRotate     = []; addParameter(p, 'thumbRotate',           d.thumbRotate,     @(x) ccTools.validators.mustBeCSSProperty(x, 'rotate'))
            case 'thumbHeight';           d.thumbHeight     = []; addParameter(p, 'thumbHeight',           d.thumbHeight,     @(x) ccTools.validators.mustBeCSSProperty(x, 'height'))
        end
    end
            
    parse(p, varargin{:});

    propStruct = struct('name', {}, 'value', {});
    propName   = setdiff(p.Parameters, p.UsingDefaults);

    for ll = 1:numel(propName)
        propValue = p.Results.(propName{ll});

        if ismember(propName{ll}, {'backgroundColor', 'backgroundHeaderColor', 'borderColor', 'color'})
            if isnumeric(p.Results.(propName{ll}))
                propValue = ccTools.fcn.rgb2hex(propValue);
            else
                propValue = char(propValue);
            end
        end

        propStruct(ll) = struct('name',  propName{ll}, ...
                                'value', propValue);
    end

end

File: +ccTools/css&js/compCustomization.js
function setup(htmlComponent) {
    htmlComponent.addEventListener("delProgressDialog", function() {
        try {
            window.parent.parent.document.getElementsByClassName("mw-busyIndicator")[0].remove();
        } catch (ME) {
            // console.log(ME)
        }
    });

    htmlComponent.addEventListener("addKeyDownListener", function(event) {
        let objDataName  = event.Data.DataName.toString();
        let objDataTag   = event.Data.DataTag.toString();
        let objHandle    = window.parent.document.querySelector(`div[data-tag="${objDataTag}"]`).children[0];

        objHandle.addEventListener('keydown', function(event) {
            keyDownListener(event, htmlComponent, objDataName, objHandle);
        });
    });

    htmlComponent.addEventListener("EditFieldFocus", function(event) {
        let objDataName  = event.Data.DataName.toString();
        let objDataTag   = event.Data.DataTag.toString();
        let objHandle    = window.parent.document.querySelector(`div[data-tag="${objDataTag}"]`).querySelector('input');

        try {
            objHandle.focus();
            objHandle.setSelectionRange(objHandle.value.length, objHandle.value.length);
        } catch (ME) {
            // console.log(ME)
        }
    });

    htmlComponent.addEventListener("htmlClassCustomization", function(event) {
        try {
            var className       = event.Data.className.toString();
            var classAttributes = event.Data.classAttributes.toString();
    
            var s = document.createElement('style');
            s.type = 'text/css';
            s.appendChild(document.createTextNode(className + " { " + classAttributes + " }"));    
            window.parent.document.head.appendChild(s);
        } catch (ME) {
            // console.log(ME)
        }
    });
        
    htmlComponent.addEventListener("compCustomization", function(event) {
        let objClass    = event.Data.Class.toString();
        let objDataTag  = event.Data.DataTag.toString();
        let objProperty = event.Data.Property.toString();
        let objValue    = event.Data.Value.toString();
        let objHandle   = window.parent.document.querySelector(`div[data-tag="${objDataTag}"]`);
        
        if (!objHandle) {
            return;
        }
        
        try {
            let elements = null;

            switch (objClass) {
                case "matlab.ui.container.ButtonGroup":
                case "matlab.ui.container.CheckBoxTree":
                case "matlab.ui.container.Panel":
                case "matlab.ui.container.Tree":
                    objHandle.style[objProperty] = objValue;
                    objHandle.children[0].style[objProperty] = objValue;
                    break;
                    
                case "matlab.ui.container.GridLayout":
                    objHandle.style.backgroundColor = objValue;
                    break;
                    
                case "matlab.ui.container.TabGroup":
                    switch (objProperty) {
                        case "backgroundColor":
                            // Pendente!
                            return;                         
                        case "backgroundHeaderColor":
                            objHandle.style.backgroundColor = "transparent";
                            objHandle.children[1].style.backgroundColor = objValue;
                            break;
                        case "transparentHeader":
                            objHandle.style.border = "none";
                            objHandle.style.backgroundColor = "transparent";
                            
                            objHandle.children[1].style.border = "none";
                            objHandle.children[1].style.backgroundColor = "transparent";                            

                            var childElements = objHandle.children[1].querySelectorAll("*");

                            childElements.forEach(function(child) {
                                child.style.border = "none";
                                child.style.backgroundColor = "transparent";                                
                            });
                            break;
                        case "borderRadius":
                        case "borderWidth":
                        case "borderColor":
                            objHandle.style[objProperty] = objValue;
                            break;
                        case "fontFamily":
                        case "fontStyle":
                        case "fontWeight":
                        case "fontSize":
                        case "color":
                            elements = objHandle.getElementsByClassName("mwTabLabel");                            
                            for (let ii = 0; ii < elements.length; ii++) {
                                elements[ii].style[objProperty] = objValue;
                            }
                    }
                    
                case "matlab.ui.container.Tab":
                    // Pendente!
                    return;
                
                case "matlab.ui.control.Button":
                case "matlab.ui.control.DropDown":
                case "matlab.ui.control.EditField":
                case "matlab.ui.control.ListBox":
                case "matlab.ui.control.NumericEditField":
                case "matlab.ui.control.StateButton":
                    objHandle.children[0].style[objProperty] = objValue;
                    break;
                    
                case "matlab.ui.control.TextArea":
                    switch (objProperty) {
                        case "backgroundColor":
                            objHandle.style.backgroundColor = "transparent";
                            objHandle.children[0].style.backgroundColor = objValue;
                            break;                            
                        case "textAlign":
                            objHandle.getElementsByTagName("textarea")[0].style.textAlign = objValue;
                            break;                            
                        default:
                            objHandle.children[0].style[objProperty] = objValue;
                    }
                    
                case "matlab.ui.control.CheckBox":
                    objHandle.getElementsByClassName("mwCheckBoxRadioIconNode")[0].style[objProperty] = objValue;
                    break;

                case "matlab.ui.control.Table":
                    switch (objProperty) {
                        case "backgroundColor":
                            objHandle.children[0].style.backgroundColor = "transparent";
                            objHandle.children[0].children[0].style.backgroundColor = objValue;
                            break;    
                        case "backgroundHeaderColor":
                            objHandle.children[0].children[0].children[0].style.backgroundColor = objValue;
                            break;    
                        case "borderRadius":
                            objHandle.children[0].style.borderRadius = objValue;
                            objHandle.children[0].children[0].style.borderRadius = objValue;
                            break;    
                        case "borderWidth":
                        case "borderColor":
                            objHandle.children[0].children[0].style[objProperty] = objValue;
                            break;
                        case "textAlign":
                        case "paddingTop":
                            elements = objHandle.getElementsByClassName("mw-table-header-row")[0].children;                      
                            for (let ii = 0; ii < elements.length; ii++) {
                                elements[ii].style[objProperty] = objValue;
                            }
                            break;
                        case "fontFamily":
                        case "fontStyle":
                        case "fontWeight":
                        case "fontSize":
                        case "color":
                            elements = objHandle.getElementsByClassName("mw-default-header-cell");
                            for (let ii = 0; ii < elements.length; ii++) {
                                elements[ii].style[objProperty] = objValue;
                            }
                    }
            }
        } catch (ME) {
            // console.log(ME)
        }
    });
}

function keyDownListener(event, htmlComponent, objDataName, objHandle) {
    keyEvents = ["ArrowUp", "ArrowDown", "Enter", "Escape"];

    switch (objDataName) {
        case "app.EditField":
            break;
        case "app.ListBox":
            let selectedItem = objHandle.getElementsByClassName("mwListItem mwListItemSelected");

            if (selectedItem.length == 1) {
                selectedItem[0].dataset.mwIndex != "1"
            }
            break;
    }

    if (keyEvents.includes(event.key)) {
        htmlComponent.sendEventToMATLAB(event.key, objDataName);
    }
}